Forget pod from scheduler's cache immediately when it's deleted

macsko opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:

When pod is in the binding phase it occupies the space on the node (is assumed). When the higher priority pod tried to preempt such pod, it deleted it from the cluster, but it remained in the scheduler's cache. Since the binding phase can take some time (PreBind, Bind), the pod occupies the space for a long time. The higher priority pod, when retried, would try to preempt the pod again, wasting scheduling cycles and API calls to the kube-apiserver.

In some scenarios, this behavior can seriously affect the scheduling performance, by continuously preempting the already deleted pod.

This PR forgets the pod from a cache when the pod delete event is registered by the scheduler. Such pod will eventually fail its binding cycle as previously, but the space won't be occupied anymore. Similar behavior is implemented for pods that are under graceful termination.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Fixes #134217

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug that caused a deleted pod staying in the binding phase to occupy space on the node in the kube-scheduler.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

macsko commented:
/test pull-kubernetes-scheduler-perf

dom4ha commented on pkg/scheduler/backend/cache/cache.go:
```diff
@@ -422,9 +422,12 @@ func (cache *cacheImpl) ForgetPod(logger klog.Logger, pod *v1.Pod) error {
 	}
-
+	if !ok {
+		// Pod is removed from the cache completely.
```
In other words, the pod wasn't assumed?

macsko replied:
Changed

dom4ha commented:
However it can help only in situation deleting pod takes immediate effect (pod is removed). Whenever removal takes time (grace period) the problem would still exist and the pod would still occupy the cache.

macsko commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -809,11 +809,11 @@ func TestAsyncPreemption(t *testing.T) {
-			name: "victim blocked in binding, preemptor pod gets activated randomly and returns to unschedulable queue until victim is bound and deleted",
+			// Victim pod takes long in binding. Preemptor pod attempts preemption, goes to unschedulable, then the victim is deleted.
+			// Preemptor pod is woken up by the Pod/Delete event and is being scheduled, even before the victim binding is terminated.
+			name: "victim blocked in binding, preemptor pod gets scheduled after victim-in-binding is deleted",
```
I had to change the test that @ania-borowiec wrote recently, because now we can schedule the preemptor pod earlier in this scenario

macsko commented:
@dom4ha @sanposhiho I updated this PR (a fix for #134217, an alternative of marking the pod as pending deletion #134205). I adjusted the integration tests, but I'm still thinking if we need more.

macsko commented:
/retest

dom4ha commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -126,12 +126,83 @@ func (sched *Scheduler) deleteNodeFromCache(obj interface{}) {
+	} else if assignedPod(newPod) {
+		// This update means binding operation. We can treat it as adding the pod to a cache
+		// (addition to the cache will handle this binding appropriately).
+		// Pod is not in the scheduling queue, so we don't have to remove it from there.
```
What if a pod that was in scheduling queue got assigned externally? Typically it should not happen, but I don't think we can assume it's our binding operation.

dom4ha commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -150,10 +221,51 @@ func (sched *Scheduler) syncPodWithDispatcher(pod *v1.Pod) *v1.Pod {
-func (sched *Scheduler) updatePodInSchedulingQueue(oldObj, newObj interface{}) {
+// handleAssumedPodDeletion is an event handler that deals with the deletion of an assumed pod.
+// We must remove it from the scheduler's cache immediately to prevent it from blocking resources for other pending pods,
+// causing unnecessary preemption attempts.
```
```suggestion
// causing unnecessary preemption attempts. Note that PreBinding/Binding will continue, but is eventually expected to fail as the pod does not exist in the api-server anymore and so in the scheduler cache.
```

macsko commented:
/retest

dom4ha commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -405,64 +472,12 @@ func addAllEventHandlers(
+		AddFunc:    sched.addPod,
+		UpdateFunc: sched.updatePod,
+		DeleteFunc: sched.deletePod,
+	}); err != nil {
```
Well, previously our deletePodFromSchedulingQueue wasn't modifying anything which should be kept between SchedulingQueue and AssignedCache, so we could have delete->add and not worry about the order in which the handlers were called.

IIUC here we need to delete cache on Delete and Update (if it changes Node or NNN assignment), but leave it on Update which is part of a binding (does not change Node assignment).

macsko replied:
Pod can't change its node

macsko commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -195,45 +316,31 @@ func hasNominatedNodeNameChanged(oldPod, newPod *v1.Pod) bool {
+		// In the unlikely event that the pod has been bound by another component,
+		// so it has to be removed from the scheduling queue to ensure correctness.
 		return
 	}
```
See that previously we didn't do that anyway. Such external binding is really unlikely and shouldn't be done when the pod is processed by the scheduler (if someone really wants to do so, they can set their custom scheduler in the pod spec). I think, relying on eventual cleanup on binding failure will be enough in this case.

macsko commented:
/retest

dom4ha reviewed (commented):
All looks good!

dom4ha commented:
/hold

dom4ha commented:
Isn't it actually better to have the eventhandler refactoring a part of Gang Scheduling PR? I worry that in an unlikely, but possible case we need to rollback, we don't have an unnecessary chain dependency.

macsko committed:
Refactor scheduler event handlers for pods to handle binding event in one place

macsko committed:
Forget pod from scheduler's cache immediately when it's deleted or has DeletionTimestamp set
